<!DOCTYPE html><html lang="en-us" >

<head>

  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="generator" content="Source Themes Academic 4.8.0">

  

  
  
  
  
  
    
    
    
  
  

  <meta name="author" content="Cassandra Buzby">

  
  
  
    
  
  <meta name="description" content="Quick recap/practice To warm up, we’re going to use one of Cassandra’s data sets, which has chromosome positions and z-scores. We’ll do each of these exercises together.
# Load in the data BSAResults = read.">

  
  <link rel="alternate" hreflang="en-us" href="https://nyusurp.github.io/2022/07/14/reproducing-visualization-in-publications/">

  


  
  
  
  <meta name="theme-color" content="#57068c">
  

  
  
  
  <script src="/js/mathjax-config.js"></script>
  

  
  
  
  
    
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/academicons/1.8.6/css/academicons.min.css" integrity="sha256-uFVgMKfistnJAfoCUQigIl+JfUaP47GrRKjf6CTPVmw=" crossorigin="anonymous">
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.0-1/css/all.min.css" integrity="sha256-4w9DunooKSr3MFXHXWyFER38WmPdm361bQS/2KUWZbU=" crossorigin="anonymous">
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.5.7/jquery.fancybox.min.css" integrity="sha256-Vzbj7sDDS/woiFS3uNKo8eIuni59rjyNGtXfstRzStA=" crossorigin="anonymous">

    
    
    
      
    
    
      
      
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.18.1/styles/github.min.css" crossorigin="anonymous" title="hl-light">
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.18.1/styles/github.min.css" crossorigin="anonymous" title="hl-dark" disabled>
      
    

    

    

    
    
      

      
      

      
    
      

      
      

      
    
      

      
      

      
    
      

      
      

      
    
      

      
      

      
    
      

      
      

      
    
      

      
      

      
    
      

      
      

      
    
      

      
      

      
    
      

      
      

      
    
      

      
      

      
        <script src="https://cdnjs.cloudflare.com/ajax/libs/lazysizes/5.1.2/lazysizes.min.js" integrity="sha256-Md1qLToewPeKjfAHU1zyPwOutccPAm5tahnaw7Osw0A=" crossorigin="anonymous" async></script>
      
    
      

      
      

      
    
      

      
      

      
    
      

      
      

      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml.js" integrity="" crossorigin="anonymous" async></script>
      
    
      

      
      

      
    

  

  
  
  
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Inconsolata:400,700%7CMerriweather%7CNanum+Myeongjo&display=swap">
  

  
  
  
  
  <link rel="stylesheet" href="/css/academic.css">

  




  


  
  

  

  <link rel="manifest" href="/index.webmanifest">
  <link rel="icon" type="image/png" href="/images/icon_hu437f07d278844891d2e703134e956aad_197157_32x32_fill_lanczos_center_2.png">
  <link rel="apple-touch-icon" type="image/png" href="/images/icon_hu437f07d278844891d2e703134e956aad_197157_192x192_fill_lanczos_center_2.png">

  <link rel="canonical" href="https://nyusurp.github.io/2022/07/14/reproducing-visualization-in-publications/">

  
  
  
  
  
  
  
    
    
  
  
  <meta property="twitter:card" content="summary">
  
  <meta property="og:site_name" content="R Workshop">
  <meta property="og:url" content="https://nyusurp.github.io/2022/07/14/reproducing-visualization-in-publications/">
  <meta property="og:title" content="Reproducing Visualization in Publications | R Workshop">
  <meta property="og:description" content="Quick recap/practice To warm up, we’re going to use one of Cassandra’s data sets, which has chromosome positions and z-scores. We’ll do each of these exercises together.
# Load in the data BSAResults = read."><meta property="og:image" content="https://nyusurp.github.io/images/icon_hu437f07d278844891d2e703134e956aad_197157_512x512_fill_lanczos_center_2.png">
  <meta property="twitter:image" content="https://nyusurp.github.io/images/icon_hu437f07d278844891d2e703134e956aad_197157_512x512_fill_lanczos_center_2.png"><meta property="og:locale" content="en-us">
  
    
      <meta property="article:published_time" content="2022-07-14T00:00:00&#43;00:00">
    
    <meta property="article:modified_time" content="2022-07-14T17:16:29-04:00">
  

  



  


  


  





  <title>Reproducing Visualization in Publications | R Workshop</title>

</head>

<body id="top" data-spy="scroll" data-offset="70" data-target="#TableOfContents" >

  <aside class="search-results" id="search">
  <div class="container">
    <section class="search-header">

      <div class="row no-gutters justify-content-between mb-3">
        <div class="col-6">
          <h1>Search</h1>
        </div>
        <div class="col-6 col-search-close">
          <a class="js-search" href="#"><i class="fas fa-times-circle text-muted" aria-hidden="true"></i></a>
        </div>
      </div>

      <div id="search-box">
        
        <input name="q" id="search-query" placeholder="Search..." autocapitalize="off"
        autocomplete="off" autocorrect="off" spellcheck="false" type="search">
        
      </div>

    </section>
    <section class="section-search-results">

      <div id="search-hits">
        
      </div>

    </section>
  </div>
</aside>


  







<nav class="navbar navbar-expand-lg navbar-light compensate-for-scrollbar" id="navbar-main">
  <div class="container">

    
    <div class="d-none d-lg-inline-flex">
      <a class="navbar-brand" href="/">R Workshop</a>
    </div>
    

    
    <button type="button" class="navbar-toggler" data-toggle="collapse"
            data-target="#navbar-content" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
    <span><i class="fas fa-bars"></i></span>
    </button>
    

    
    <div class="navbar-brand-mobile-wrapper d-inline-flex d-lg-none">
      <a class="navbar-brand" href="/">R Workshop</a>
    </div>
    

    
    
    <div class="navbar-collapse main-menu-item collapse justify-content-start" id="navbar-content">

      
      <ul class="navbar-nav d-md-inline-flex">
        

        
        <li class="nav-item dropdown">
          <a href="#" class="nav-link dropdown-toggle" data-toggle="dropdown" aria-haspopup="true"><span>Material</span><span class="caret"></span>
          </a>
          <div class="dropdown-menu">
            
              <a class="dropdown-item" href="/#handout"><span>Handouts</span></a>
            
              <a class="dropdown-item" href="/extra/"><span>Extra</span></a>
            
          </div>
        </li>

        
        

        
        <li class="nav-item dropdown">
          <a href="#" class="nav-link dropdown-toggle" data-toggle="dropdown" aria-haspopup="true"><span>Worksheet and Data</span><span class="caret"></span>
          </a>
          <div class="dropdown-menu">
            
              <a class="dropdown-item" href="/assets/session1.zip"><span>Session 1</span></a>
            
              <a class="dropdown-item" href="/assets/session2.zip"><span>Session 2</span></a>
            
              <a class="dropdown-item" href="/assets/session3.zip"><span>Session 3</span></a>
            
              <a class="dropdown-item" href="/assets/session4.zip"><span>Session 4</span></a>
            
              <a class="dropdown-item" href="/assets/session5.zip"><span>Session 5</span></a>
            
              <a class="dropdown-item" href="/assets/session6.zip"><span>Session 6</span></a>
            
              <a class="dropdown-item" href="/assets/session7.zip"><span>Session 7</span></a>
            
              <a class="dropdown-item" href="/assets/session8.zip"><span>Session 8</span></a>
            
              <a class="dropdown-item" href="/assets/session9.zip"><span>Session 9</span></a>
            
              <a class="dropdown-item" href="/assets/session10.zip"><span>Session 10</span></a>
            
          </div>
        </li>

        
        

        
        <li class="nav-item dropdown">
          <a href="#" class="nav-link dropdown-toggle" data-toggle="dropdown" aria-haspopup="true"><span>Slide Deck</span><span class="caret"></span>
          </a>
          <div class="dropdown-menu">
            
              <a class="dropdown-item" href="/slides/session_1/session_1_welcome.html"><span>Session 1: Welcome</span></a>
            
              <a class="dropdown-item" href="/slides/session_2/session_2_EDA_summary.html"><span>Session 2: dplyr</span></a>
            
              <a class="dropdown-item" href="/slides/session_6/function_and_data_types.html"><span>Session 6: Functions and Data Types</span></a>
            
          </div>
        </li>

        
        

      

        
      </ul>
    </div>

    <ul class="nav-icons navbar-nav flex-row ml-auto d-flex pl-md-2">
      
      <li class="nav-item">
        <a class="nav-link js-search" href="#" aria-label="Search"><i class="fas fa-search" aria-hidden="true"></i></a>
      </li>
      

      

      

    </ul>

  </div>
</nav>


  <article class="article">

  












  

  
  
  
<div class="article-container pt-3">
  <h1>Reproducing Visualization in Publications</h1>

  

  
    


<div class="article-metadata">

  
  
  
  
  <div>
    

  
  <span><a href="/author/yen-chung-chen/">Yen-Chung Chen</a></span>
  </div>
  
  

  
  <span class="article-date">
    
    
      
          Last updated on
      
    
    2022-07-14
  </span>
  

  

  

  
  
  

  
  
  <span class="middot-divider"></span>
  <span class="article-categories">
    <i class="fas fa-folder mr-1"></i><a href="/category/handout/">Handout</a></span>
  

</div>

    














  
</div>



  <div class="article-container">

    <div class="article-style">
      


<div id="quick-recappractice" class="section level2">
<h2>Quick recap/practice</h2>
<p>To warm up, we’re going to use one of Cassandra’s data sets, which has chromosome positions and z-scores. We’ll do each of these exercises together.</p>
<pre class="r"><code># Load in the data
BSAResults = read.csv(&quot;data/BSAResults.csv&quot;)

# Find the unique values of the chromosomes. How do we find the column name?
str(BSAResults)</code></pre>
<pre><code>## &#39;data.frame&#39;:    49432 obs. of  6 variables:
##  $ X          : chr  &quot;result.1&quot; &quot;result.2&quot; &quot;result.3&quot; &quot;result.4&quot; ...
##  $ CHROM      : chr  &quot;I&quot; &quot;I&quot; &quot;I&quot; &quot;I&quot; ...
##  $ POS        : num  100007 1035 1074 1101 1173 ...
##  $ Bulk       : num  -0.08758 0.1827 -0.00987 0.01895 0.87397 ...
##  $ Parent     : num  3.554 0.537 -0.345 -0.12 -0.393 ...
##  $ Interaction: num  -0.01715 0.13848 0.31904 -0.00762 -0.18258 ...</code></pre>
<pre class="r"><code>unique(BSAResults$CHROM)</code></pre>
<pre><code>##  [1] &quot;I&quot;    &quot;II&quot;   &quot;III&quot;  &quot;IV&quot;   &quot;V&quot;    &quot;VI&quot;   &quot;VII&quot;  &quot;VIII&quot; &quot;IX&quot;   &quot;X&quot;   
## [11] &quot;XI&quot;   &quot;XII&quot;  &quot;XIII&quot; &quot;XIV&quot;  &quot;XV&quot;   &quot;XVI&quot;</code></pre>
<pre class="r"><code># Filter for Chromosome II using dplyr&#39;s filter() function, and pipe through head() to output only the first 10 lines
BSAResults %&gt;%
  filter(CHROM == &quot;II&quot;) %&gt;%
  head(., n = 10)</code></pre>
<pre><code>##               X CHROM    POS        Bulk        Parent Interaction
## 1  result.11000    II 100153 -0.18330726 -0.0646732103  0.19943965
## 2  result.21000    II 100367 -0.18800155 -0.2912844665  0.28590390
## 3  result.31000    II 100413 -0.32764216 -0.3031255854  0.52362311
## 4  result.41000    II 100878  0.02040887  0.1185267276 -0.16398910
## 5  result.51000    II 101103 -0.13385976 -0.0100344121 -0.01602947
## 6   result.6432    II 101289 -0.25282081  0.0817157785  0.21342401
## 7   result.7100    II 101340 -0.19522650  0.1584760726 -0.15293462
## 8   result.8100    II  10136 -0.30926532 -0.0078019497 -0.02303257
## 9   result.9100    II 101507 -0.19850031 -0.0007044734  0.08039453
## 10 result.10100    II 101581  0.07333291  0.0641751118 -0.28421660</code></pre>
<pre class="r"><code># Plot a scatter plot of the Bulk with x-axis being POS and y-axis being Bulk values
BSAResults %&gt;%
  filter(CHROM == &quot;II&quot;) %&gt;%
  ggplot(., aes(x = POS, y = Bulk)) +
  geom_point()</code></pre>
<p><img src="https://nyusurp.github.io/2022/07/14/reproducing-visualization-in-publications/index.en_files/figure-html/unnamed-chunk-1-1.png" width="960" /></p>
<pre class="r"><code># If we want plot Bulk, Interaction, and Parent for Chr II at once, we need to pivot the data to long format. How do we do so?
BSAResults %&gt;%
  pivot_longer(., cols = c(Bulk, Interaction, Parent)) %&gt;%
  head()</code></pre>
<pre><code>## # A tibble: 6 × 5
##   X        CHROM    POS name          value
##   &lt;chr&gt;    &lt;chr&gt;  &lt;dbl&gt; &lt;chr&gt;         &lt;dbl&gt;
## 1 result.1 I     100007 Bulk        -0.0876
## 2 result.1 I     100007 Interaction -0.0171
## 3 result.1 I     100007 Parent       3.55  
## 4 result.2 I       1035 Bulk         0.183 
## 5 result.2 I       1035 Interaction  0.138 
## 6 result.2 I       1035 Parent       0.537</code></pre>
<pre class="r"><code># Then to plot, we can separate our data by &quot;name&quot; - rename this column &quot;Z_score&quot; using pivot_longer()
BSAResults %&gt;%
  pivot_longer(., cols = c(Bulk, Interaction, Parent), names_to = &quot;Z_score&quot;) %&gt;%
# There&#39;s a more interesting thing happening on Chr VIII. Filter for this instead.
  filter(CHROM == &quot;VIII&quot;) %&gt;%
# Plot the pivoted points with them colored by Z_score. 
  ggplot(., aes(x = POS, y = value, color = Z_score)) +
  geom_point()</code></pre>
<p><img src="https://nyusurp.github.io/2022/07/14/reproducing-visualization-in-publications/index.en_files/figure-html/unnamed-chunk-1-2.png" width="960" /></p>
<pre class="r"><code># There are a lot of points. Change the opacity of these points.
BSAResults %&gt;%
  pivot_longer(., cols = c(Bulk, Interaction, Parent), names_to = &quot;Z_score&quot;) %&gt;%
  filter(CHROM == &quot;VIII&quot;) %&gt;%
  ggplot(., aes(x = POS, y = value, color = Z_score)) +
  geom_point(alpha = 0.3)</code></pre>
<p><img src="https://nyusurp.github.io/2022/07/14/reproducing-visualization-in-publications/index.en_files/figure-html/unnamed-chunk-1-3.png" width="960" /></p>
<pre class="r"><code># We now want to change the names to reflect what the value actually is. Add labels to the plot.
BSAResults %&gt;%
  pivot_longer(., cols = c(Bulk, Interaction, Parent), names_to = &quot;Z_score&quot;) %&gt;%
  filter(CHROM == &quot;VIII&quot;) %&gt;%
  ggplot(., aes(x = POS, y = value, color = Z_score)) +
  geom_point(alpha = 0.3) +
  xlab(&quot;Coordinate on Chromsome (nt)&quot;) +
  ylab(&quot;Z-Score&quot;) +
  ggtitle(&quot;The Light from the Other End of the Tunnel&quot;) +
# A different theme might be better. Let&#39;s use theme_classic()
  theme_classic()</code></pre>
<p><img src="https://nyusurp.github.io/2022/07/14/reproducing-visualization-in-publications/index.en_files/figure-html/unnamed-chunk-1-4.png" width="960" /></p>
<pre class="r"><code># BONUS: it would be great to plot all chromosomes on the plot at once. Add facet_grid(~CHROM) to do this, and once it works, remove the filtering of your CHROM column.
BSAResults %&gt;%
  pivot_longer(., cols = c(Bulk, Interaction, Parent), names_to = &quot;Z_score&quot;) %&gt;%
  # filter(CHROM == &quot;VIII&quot;) %&gt;%
  ggplot(., aes(x = POS, y = value, color = Z_score)) +
  geom_point(alpha = 0.3, size = 0.1) +
  xlab(&quot;Coordinate on Chromsome (nt)&quot;) +
  ylab(&quot;Z-Score&quot;) +
  ggtitle(&quot;The One where Everything is Plotted Together&quot;) +
  theme_classic() +
  facet_grid(~CHROM, space = &quot;free_x&quot;) +
  # NEW: Another way to limit the y
  coord_cartesian(ylim = c(-1.5, 1.5)) +
  guides(color = &quot;none&quot;)</code></pre>
<p><img src="https://nyusurp.github.io/2022/07/14/reproducing-visualization-in-publications/index.en_files/figure-html/unnamed-chunk-1-5.png" width="960" /></p>
<p>Today, we will be looking into the first figure from
<a href="https://www.nature.com/articles/s41587-020-0456-9">Massively parallel Cas13 screens reveal principles for guide RNA design</a> from the Sanjana Lab.</p>
<p>The main reason we selected this paper is because: First, the visualization they
used are fairly popular and meets what we covered in previous sessions; second,
they did an excellent job in
<a href="https://gitlab.com/sanjanalab/cas13">sharing their data and code</a>, so you will
have this unique chance to work with real data and know exactly how the authors
performed their analyses and visualization if you take the time to read their
code.</p>
</div>
<div id="figure-1b-boxplot" class="section level2">
<h2>Figure 1b: BoxPlot</h2>
<p>Figure 1b is a panel of boxplots showing log-transformed enrichment scores of
different types of crRNAs (the equivalent to sgRNAs for Cas13a).</p>
<p>The processed (normalized and batch-corrected) data for crRNAs in this tiling
screen is provided in <code>data/GFP_screen_crRNA_enrichments.csv</code>. Specifically:</p>
<ul>
<li>crRNAs are categorized in to several categories based on their targets in
<code>type</code></li>
<li>Log-transformed enrichment scores are stored in <code>meanCS.BIN1</code></li>
</ul>
<p>First, let’s load in the data and see what it looks like</p>
<pre class="r"><code>#Load in our data
GFP_Results = read.csv(&quot;data/GFP_screen_crRNA_enrichments.csv&quot;)

#See what the data looks like
str(GFP_Results)</code></pre>
<pre><code>## &#39;data.frame&#39;:    7500 obs. of  17 variables:
##  $ p.BIN1       : num  0.03714 0.00531 0.05278 0.00993 0.81008 ...
##  $ p.BIN2       : num  0.00441 0.06989 1 0.31136 1 ...
##  $ p.BIN3       : num  0.0164 0.038 0.3089 0.2385 0.6272 ...
##  $ p.BIN4       : num  0.1162 0.0542 0.2063 0.012 0.8689 ...
##  $ logBIN1      : num  1.4302 2.2752 1.2776 2.0029 0.0915 ...
##  $ logBIN2      : num  2.355 1.156 0 0.507 0 ...
##  $ logBIN3      : num  1.785 1.42 0.51 0.622 0.203 ...
##  $ logBIN4      : num  0.935 1.266 0.686 1.923 0.061 ...
##  $ meanCS.BIN1  : num  0.698 0.973 1.167 1.055 0.953 ...
##  $ meanCS.BIN2  : num  -0.6699 -0.2926 -0.0292 -0.3766 NA ...
##  $ meanCS.BIN3  : num  -0.923 -0.469 -0.589 -0.499 -0.852 ...
##  $ meanCS.BIN4  : num  -0.658 -0.408 -0.573 -0.689 -0.67 ...
##  $ medianCS.BIN1: num  0.742 0.97 1.167 1.235 0.953 ...
##  $ medianCS.BIN2: num  -0.6353 -0.2197 -0.0292 -0.0923 NA ...
##  $ medianCS.BIN3: num  -1.019 -0.446 -0.589 -0.458 -0.852 ...
##  $ medianCS.BIN4: num  -0.774 -0.398 -0.573 -0.762 -0.67 ...
##  $ type         : chr  &quot;Perfect Match&quot; &quot;Perfect Match&quot; &quot;Perfect Match&quot; &quot;Perfect Match&quot; ...</code></pre>
<p>Next we want to change the “type” column into factors. We can do this by using the function factor() which we went over last class, and then add the labels function to change the words into acronyms.</p>
<pre class="r"><code># Note that each type is:
# 1. Represented as acronyms in the figure
# 2. Are NOT arranged in alphabetical order
# We can see the need to transform the type column to a factor

GFP_Results$type = factor(
  x = GFP_Results$type,
  levels = c(
    &quot;Perfect Match&quot;,
    &quot;First Order&quot;,
    &quot;Random Double&quot;,
    &quot;Consecutive Double&quot;,
    &quot;Consecutive Triple&quot;,
    &quot;Non-Targeting&quot;
  ),
  labels = c(
    &quot;PM&quot;,
    &quot;SM&quot;,
    &quot;RD&quot;, 
    &quot;CD&quot;,
    &quot;CT&quot;,
    &quot;NT&quot;
  )
)</code></pre>
<p>Next we can make a basic boxplot, using ggplot and filtering for non NA values:</p>
<pre class="r"><code>GFP_Results %&gt;%
  # This is to remove the rows where meanCS.BIN is NA
  filter(!is.na(meanCS.BIN1)) %&gt;%
  #basic elements of the plot
  ggplot(., aes(x = type, y = meanCS.BIN1)) +
  geom_boxplot() +
  #labels
  ggtitle(&quot;All crRNAs&quot;) +
  xlab(&quot;&quot;) +
  ylab(&quot;log2(FC)&quot;)</code></pre>
<p><img src="https://nyusurp.github.io/2022/07/14/reproducing-visualization-in-publications/index.en_files/figure-html/unnamed-chunk-4-1.png" width="672" /></p>
<p>First of all, the colors are different. We can change these by setting the hex codes to match the Sanjana lab’s colors, then inputting this vector of hex codes into the scale_fill_manual() function:</p>
<pre class="r"><code># The color palette used in the paper
#### BONUS: Change the colors and run the chunk to see what happens to the
#### plot!!
sanjana_colors = c(
  &quot;#E69F00&quot;, &quot;#D55E00&quot;, &quot;#009E73&quot;, &quot;#0072B2&quot;, &quot;#56B4E9&quot;, &quot;#999999&quot;
)

GFP_Results %&gt;%
  filter(!is.na(meanCS.BIN1)) %&gt;%
  #basic elements of the plot
  ggplot(aes(fill = type, x = type, y = meanCS.BIN1)) + geom_boxplot() +
  #labels
  xlab(&quot;&quot;) + ylab(&quot;log2(FC)&quot;) + ggtitle(&quot;All crRNAs&quot;) +
  #colors
  scale_fill_manual(values = sanjana_colors)</code></pre>
<p><img src="https://nyusurp.github.io/2022/07/14/reproducing-visualization-in-publications/index.en_files/figure-html/unnamed-chunk-5-1.png" width="672" /></p>
<p>We can also make the outliers smaller, using the outlier.shape and outlier.size functions in geom_boxplot().</p>
<pre class="r"><code>GFP_Results %&gt;%
  filter(!is.na(meanCS.BIN1)) %&gt;%
  #basic elements of the plot
  ggplot(aes(fill = type, x = type, y = meanCS.BIN1)) + geom_boxplot(outlier.shape = 20, outlier.size = 0.05) +
  #labels
  xlab(&quot;&quot;) + ylab(&quot;log2(FC)&quot;) + ggtitle(&quot;All crRNAs&quot;) +
  #colors
  scale_fill_manual(values = sanjana_colors)</code></pre>
<p><img src="https://nyusurp.github.io/2022/07/14/reproducing-visualization-in-publications/index.en_files/figure-html/unnamed-chunk-6-1.png" width="672" />
Next, let’s reduce the white space by adding a ylim argument:</p>
<pre class="r"><code>GFP_Results %&gt;%
  filter(!is.na(meanCS.BIN1)) %&gt;%
  #basic elements of the plot
  ggplot(aes(fill = type, x = type, y = meanCS.BIN1)) + geom_boxplot(outlier.shape = 20, outlier.size = 0.05) +
  #labels
  xlab(&quot;&quot;) + ylab(&quot;log2(FC)&quot;) + ggtitle(&quot;All crRNAs&quot;) +
  #colors
  scale_fill_manual(values = sanjana_colors) +
  #y axis limits
  ylim(c(-1.5,1.5)) </code></pre>
<pre><code>## Warning: Removed 15 rows containing non-finite values (stat_boxplot).</code></pre>
<p><img src="https://nyusurp.github.io/2022/07/14/reproducing-visualization-in-publications/index.en_files/figure-html/unnamed-chunk-7-1.png" width="672" /></p>
<p>Finally, we’re going to use a new function called coord_fixed() to restrict the aspect ratio of the plot.</p>
<pre class="r"><code>GFP_Results %&gt;%
  filter(!is.na(meanCS.BIN1)) %&gt;%
  #basic elements of the plot
  ggplot(aes(fill = type, x = type, y = meanCS.BIN1)) + geom_boxplot(outlier.shape = 20, outlier.size = 0.05) +
  #labels
  xlab(&quot;&quot;) + ylab(&quot;log2(FC)&quot;) + ggtitle(&quot;All crRNAs&quot;) +
  #colors
  scale_fill_manual(values = sanjana_colors) +
  #y-axis limits and ratio
  ylim(c(-1.5,1.5)) + coord_fixed(ratio = 6/3) +
  # Remove legends
    guides(fill = &quot;none&quot;) +
    # Remove x-axis line and tick
    theme(
      axis.ticks.x = element_blank(), 
      axis.line.x = element_blank()
    )</code></pre>
<pre><code>## Warning: Removed 15 rows containing non-finite values (stat_boxplot).</code></pre>
<p><img src="https://nyusurp.github.io/2022/07/14/reproducing-visualization-in-publications/index.en_files/figure-html/unnamed-chunk-8-1.png" width="672" /></p>
<p>And finally, for a publication-level plot, we’ll remove our legend and x axis ticks using guides() and theme(), and then set our overall theme to theme_classic().</p>
<pre class="r"><code>GFP_Results %&gt;%
  filter(!is.na(meanCS.BIN1)) %&gt;%
  #basic elements of the plot
  ggplot(aes(fill = type, x = type, y = meanCS.BIN1)) + geom_boxplot(outlier.shape = 20, outlier.size = 0.05) +
  #labels
  xlab(&quot;&quot;) + ylab(&quot;log2(FC)&quot;) + ggtitle(&quot;All crRNAs&quot;) +
  #colors
  scale_fill_manual(values = sanjana_colors) +
  #y axis limits and ratio
  ylim(c(-1.5,1.5)) + coord_fixed(ratio = 6/3) +
  # Remove legends
  guides(fill = &quot;none&quot;) +
  theme_classic() +
  # Remove x-axis line and tick
    theme(
      axis.ticks.x = element_blank(), 
      axis.line.x = element_blank()
    )</code></pre>
<pre><code>## Warning: Removed 15 rows containing non-finite values (stat_boxplot).</code></pre>
<p><img src="https://nyusurp.github.io/2022/07/14/reproducing-visualization-in-publications/index.en_files/figure-html/unnamed-chunk-9-1.png" width="672" /></p>
<div id="section" class="section level79">
<p class="heading"></p>
</div>
</div>
<div id="figure-1c-scatter-plot-with-jitter" class="section level2">
<h2>Figure 1c: Scatter plot with jitter</h2>
<p>Figure 1c is a panel of scatter plots with jitter showing differences in
log-transformed enrichment scores of individual crRNAs of different types.</p>
<p>The processed data for individual crRNAs is provided in
<code>data/GFPTiling_individual.csv</code>. Specifically:</p>
<ul>
<li>crRNAs are categorized in to several categories based on their targets in
<code>MatchType</code></li>
<li>Differences of log-transformed enrichment scores are stored in <code>DeltaCS</code></li>
</ul>
<pre class="r"><code># Load data
tiling_result = read.csv(&quot;data/GFPTiling_individual.csv&quot;)

# Take a look at your data
str(tiling_result)</code></pre>
<pre><code>## &#39;data.frame&#39;:    179 obs. of  10 variables:
##  $ Screen    : chr  &quot;GFP&quot; &quot;GFP&quot; &quot;GFP&quot; &quot;GFP&quot; ...
##  $ GuideName : chr  &quot;crRNA004:7-33&quot; &quot;crRNA004:7-33_consecDouble_01&quot; &quot;crRNA004:7-33_consecDouble_02&quot; &quot;crRNA004:7-33_consecDouble_03&quot; ...
##  $ Guide     : chr  &quot;crRNA004&quot; &quot;crRNA004&quot; &quot;crRNA004&quot; &quot;crRNA004&quot; ...
##  $ MatchType : chr  &quot;Perfect Match&quot; &quot;Consecutive Double&quot; &quot;Consecutive Double&quot; &quot;Consecutive Double&quot; ...
##  $ MatchPos  : int  33 33 33 33 33 33 33 33 33 33 ...
##  $ Annotation: chr  &quot;CDS&quot; &quot;CDS&quot; &quot;CDS&quot; &quot;CDS&quot; ...
##  $ meanCS    : num  1.0546 0.9534 0.839 0.4866 -0.0992 ...
##  $ pVal      : num  0.00993 0.81008 0.02679 0.07028 0.172 ...
##  $ log10pVal : num  2.0029 0.0915 1.572 1.1532 0.7645 ...
##  $ DeltaCS   : num  0 -0.101 -0.216 -0.568 -1.154 ...</code></pre>
<pre class="r"><code>#Find the unique values for your MatchType column
unique(tiling_result$MatchType)</code></pre>
<pre><code>## [1] &quot;Perfect Match&quot;      &quot;Consecutive Double&quot; &quot;Consecutive Triple&quot;
## [4] &quot;First Order&quot;        &quot;Random Double&quot;</code></pre>
<pre class="r"><code># Make MatchType a factor
tiling_result$MatchType = factor(
  x = tiling_result$MatchType,
  levels = c(
    &quot;Perfect Match&quot;,
    &quot;First Order&quot;,
    &quot;Random Double&quot;,
    &quot;Consecutive Double&quot;,
    &quot;Consecutive Triple&quot;
  ),
  labels = c(
    &quot;PM&quot;,
    &quot;SM&quot;,
    &quot;RD&quot;, 
    &quot;CD&quot;,
    &quot;CT&quot;
  )
)</code></pre>
<p>Next, let’s make the basic ggplot for the tiling result, with MatchType on the x axis, DeltaCS on the y axis, and colored points by MatchType. You’ll want to use geom_jitter() rather than geom_point(), with a width of 0.2, a size of 3, and a shape of 20.</p>
<pre class="r"><code>tiling_result %&gt;%
#Filter out the NAs in the DeltaCS column
  filter(!is.na(DeltaCS)) %&gt;%
#Make your ggplot scatter plot here
  ggplot(., aes(x = MatchType, y = DeltaCS, color = MatchType)) +
  geom_jitter()</code></pre>
<p><img src="https://nyusurp.github.io/2022/07/14/reproducing-visualization-in-publications/index.en_files/figure-html/unnamed-chunk-11-1.png" width="672" /></p>
<p>We’ll add horizontal lines using the new function stat_summary(), which provides plotting of simple summary statistics. First, let’s look up the documentation for stat_summary().</p>
<pre class="r"><code>#What is the other way to find the help menu?
?stat_summary</code></pre>
<p>Let’s go ahead and use stat_summary, finding the median of each group:</p>
<pre class="r"><code>tiling_result %&gt;%
  filter(!is.na(DeltaCS)) %&gt;%
  ggplot(aes(x = MatchType, y = DeltaCS, color = MatchType)) +
    geom_jitter(
      # Do you notice several things are different from your plot above
      # and the real one in the paper?
      # The size of the points and how wide they spread/jitter are the most
      # obvious to me, so we are fixing those here.
      width = 0.2,size = 3,shape = 20
    ) +
  
    # Add a horizontal bar at the median of each group
    # NEW FUNCTION!! stat_summary() provides plotting of simple summary
    # statistics (defined by fun (function to summarize each group)) with several
    # types of plots (defined by geom)
    stat_summary(
      fun = median,
      geom = &quot;crossbar&quot;,
      width = 0.5,
      color = &quot;black&quot;
    ) </code></pre>
<p><img src="https://nyusurp.github.io/2022/07/14/reproducing-visualization-in-publications/index.en_files/figure-html/unnamed-chunk-13-1.png" width="672" /></p>
<p>Next, we can scale our y axis continuously, as follows:</p>
<pre class="r"><code>tiling_result %&gt;%
  filter(!is.na(DeltaCS)) %&gt;%
  ggplot(aes(x = MatchType, y = DeltaCS, color = MatchType)) +
    geom_jitter(width = 0.2,size = 3,shape = 20) +
    stat_summary(fun = median,geom = &quot;crossbar&quot;,width = 0.5, color = &quot;black&quot;) + 
  
  #scaling function
    scale_y_continuous(limits = c(-2, 0.5), breaks = c(seq(-2, 0)))</code></pre>
<pre><code>## Warning: Removed 2 rows containing non-finite values (stat_summary).</code></pre>
<pre><code>## Warning: Removed 2 rows containing missing values (geom_point).</code></pre>
<p><img src="https://nyusurp.github.io/2022/07/14/reproducing-visualization-in-publications/index.en_files/figure-html/unnamed-chunk-14-1.png" width="672" /></p>
<p>To finish up this plot, we’ll want to set the following arguments. With a partner, add these to your plot:</p>
<ul>
<li>title your plot “Individual crRNA”</li>
<li>remove the x axis label</li>
<li>set the y axis label to “delta log2(FC)”</li>
<li>set the coordinate ratio to 7/3.5</li>
<li>set the colors to sanajana colors as we did before (note that this time the function has to be scale_color_manual rather than scale_fill_manual)</li>
<li>Remove x-axis line and tick as we did before</li>
</ul>
<pre class="r"><code># Your publication-ready plot here
tiling_result %&gt;%
  filter(!is.na(DeltaCS)) %&gt;%
  ggplot(aes(x = MatchType, y = DeltaCS, color = MatchType)) +
  geom_jitter(width = 0.2,size = 3,shape = 20) +
  stat_summary(fun = median,geom = &quot;crossbar&quot;,width = 0.5, color = &quot;black&quot;) + 
  
  #scaling function
  scale_y_continuous(limits = c(-2, 0.5), breaks = c(seq(-2, 0))) +
  scale_color_manual(values = sanjana_colors) +
  coord_fixed(ratio = 7/3.5) +
  # Set axis labels
  xlab(&quot;&quot;) +
  ylab(&quot;Delta log2(FC)&quot;) +
  ggtitle(&quot;Individual crRNAs&quot;) +
  theme_classic() +
  theme(
    axis.line.x = element_blank(),
    axis.ticks.x = element_blank()
  ) +
  guides(color = &quot;none&quot;)</code></pre>
<pre><code>## Warning: Removed 2 rows containing non-finite values (stat_summary).</code></pre>
<pre><code>## Warning: Removed 2 rows containing missing values (geom_point).</code></pre>
<p><img src="https://nyusurp.github.io/2022/07/14/reproducing-visualization-in-publications/index.en_files/figure-html/unnamed-chunk-15-1.png" width="672" /></p>
<div id="section-1" class="section level79">
<p class="heading"></p>
</div>
</div>
<div id="figure-1d-scatter-plot-with-a-smooth-line" class="section level2">
<h2>Figure 1d: Scatter plot with a smooth line</h2>
<p>Figure 1d is a panel of scatter plots with a smooth line for enrichment scores
of individual crRNAs targeting different regions of the GFP transcript.</p>
<p>The processed data for individual crRNAs that target GFP is provided in
<code>data/GFPTiling_individual_full.csv</code>. Specifically:</p>
<ul>
<li>Log-transformed enrichment scores are stored in <code>meanCS</code></li>
<li>The points are plotted over the GFP transcripts, and the coordinates of
matching site is in <code>MatchPos</code></li>
<li>The points were colored by which quantile it falls into.</li>
</ul>
<p>Quantile is just binning values into four bins, and <code>dplyr</code> has another
function to determine *which quantile a row belongs to by a numeric column
of your choice, and the syntax is <code>ntile(column, number_of_bins)</code>.</p>
<p>A minimal example of its usage is:</p>
<pre class="r"><code>exdf = data.frame(
  value = 1:10
)


exdf %&gt;% mutate(
  split_at_median = ntile(value, 2),
  quantile = ntile(value, 4)
)</code></pre>
<pre><code>##    value split_at_median quantile
## 1      1               1        1
## 2      2               1        1
## 3      3               1        1
## 4      4               1        2
## 5      5               1        2
## 6      6               2        2
## 7      7               2        3
## 8      8               2        3
## 9      9               2        4
## 10    10               2        4</code></pre>
<p>Please observe the figure and propose your way of reproducing the panel. The
colors they used are provided in the chunk below:</p>
<pre class="r"><code>gfp_color = c(&#39;#ca0020&#39;,&#39;#f4a582&#39;,&#39;#92c5de&#39;,&#39;#0571b0&#39;)</code></pre>
<pre class="r"><code># Load data
allGFP = read.csv(&quot;data/GFPTiling_individual_full.csv&quot;)</code></pre>
<pre class="r"><code># Use qtile() within mutate to create our data frame
allGFP %&gt;%
  filter(!is.na(meanCS)) %&gt;%
  # Assign quartile (split into 4 proportions) by meanCS
  # Since we are going to color by quartile, it is preferable to use factor
  # than numerics
  mutate(qtile = as.factor(ntile(meanCS, 4))) %&gt;% head(20)</code></pre>
<pre><code>##    Screen      GuideName    Guide     MatchType MatchPos Annotation
## 1     GFP  crRNA001:1-27 crRNA001 Perfect Match       27        CDS
## 2     GFP  crRNA002:2-28 crRNA002 Perfect Match       28        CDS
## 3     GFP  crRNA003:6-32 crRNA003 Perfect Match       32        CDS
## 4     GFP  crRNA004:7-33 crRNA004 Perfect Match       33        CDS
## 5     GFP  crRNA005:9-35 crRNA005 Perfect Match       35        CDS
## 6     GFP crRNA006:10-36 crRNA006 Perfect Match       36        CDS
## 7     GFP crRNA007:11-37 crRNA007 Perfect Match       37        CDS
## 8     GFP crRNA008:12-38 crRNA008 Perfect Match       38        CDS
## 9     GFP crRNA009:14-40 crRNA009 Perfect Match       40        CDS
## 10    GFP crRNA010:15-41 crRNA010 Perfect Match       41        CDS
## 11    GFP crRNA011:16-42 crRNA011 Perfect Match       42        CDS
## 12    GFP crRNA012:20-46 crRNA012 Perfect Match       46        CDS
## 13    GFP crRNA013:21-47 crRNA013 Perfect Match       47        CDS
## 14    GFP crRNA014:22-48 crRNA014 Perfect Match       48        CDS
## 15    GFP crRNA015:26-52 crRNA015 Perfect Match       52        CDS
## 16    GFP crRNA016:27-53 crRNA016 Perfect Match       53        CDS
## 17    GFP crRNA017:30-56 crRNA017 Perfect Match       56        CDS
## 18    GFP crRNA018:31-57 crRNA018 Perfect Match       57        CDS
## 19    GFP crRNA019:33-59 crRNA019 Perfect Match       59        CDS
## 20    GFP crRNA020:37-63 crRNA020 Perfect Match       63        CDS
##          meanCS        pVal log10pVal    ScaledCS qtile
## 1   0.697828662 0.037139487 1.4301641  1.08371271     3
## 2   0.973333658 0.005305903 2.2752407  1.62662808     4
## 3   1.166888147 0.052775054 1.2775713  2.00805023     4
## 4   1.054580427 0.009933042 2.0029177  1.78673451     4
## 5   0.850456714 0.161383678 0.7921404  1.38448445     4
## 6   0.957187296 0.013888889 1.8573325  1.59480976     4
## 7   0.985674966 0.136754065 0.8640598  1.65094810     4
## 8   1.074104147 0.555658488 0.2551920  1.82520832     4
## 9  -0.080430937 0.600709794 0.2213353 -0.44994042     1
## 10  0.786329262 0.015443537 1.8112532  1.25811367     3
## 11 -0.068862813 0.365481687 0.4371344 -0.42714406     1
## 12  0.047882725 0.597939473 0.2233428 -0.19708308     1
## 13  0.725672983 0.168281759 0.7739630  1.13858326     3
## 14  0.108779029 0.434274771 0.3622354 -0.07707967     2
## 15 -0.038672492 0.858834132 0.0660907 -0.36765044     1
## 16 -0.056426383 0.295065651 0.5300813 -0.40263659     1
## 17  0.132830725 0.552465793 0.2576946 -0.02968294     2
## 18  0.009357908 0.763074183 0.1174332 -0.27300082     1
## 19 -0.043976387 0.754415576 0.1223894 -0.37810239     1
## 20  0.519437522 0.137494396 0.8617150  0.73217175     3</code></pre>
<pre class="r"><code> allGFP %&gt;%
  filter(!is.na(meanCS)) %&gt;%
  mutate(qtile = as.factor(ntile(meanCS, 4))) %&gt;%
  # Pipe into ggplot and define the columns used for axes and color
  ggplot(aes(x = MatchPos, y = meanCS, color = qtile)) +
  # Plot a smooth line by the scatter points you provided 
  geom_smooth(
    # There are multiple ways to get a line from points
    # Here, we use &quot;locally estimated scatter plot smoothing&quot;
    method = &quot;loess&quot;,
    # which fits a polynomial function (curve line) within a small window
    # whose size is defined by span
    span = 0.05,
    # The color of the smooth line
    color = &quot;black&quot;,
    # The size of the smooth line
    size = 0.5,
    # The color of the shaded area representing the standard error of
    # the smoothing method you picked
    fill = &quot;grey77&quot;
  ) </code></pre>
<pre><code>## `geom_smooth()` using formula &#39;y ~ x&#39;</code></pre>
<p><img src="https://nyusurp.github.io/2022/07/14/reproducing-visualization-in-publications/index.en_files/figure-html/unnamed-chunk-20-1.png" width="672" /></p>
<p>Then of course we can add the points, colored by their qtile scores</p>
<pre class="r"><code>allGFP %&gt;%
  filter(!is.na(meanCS)) %&gt;%
  mutate(qtile = as.factor(ntile(meanCS, 4))) %&gt;%
  ggplot(aes(x = MatchPos, y = meanCS, color = qtile)) +
  geom_smooth(method = &quot;loess&quot;, span = 0.05, color = &quot;black&quot;, size = 0.5, fill = &quot;grey77&quot;) +
  geom_point(shape = 20)</code></pre>
<pre><code>## `geom_smooth()` using formula &#39;y ~ x&#39;</code></pre>
<p><img src="https://nyusurp.github.io/2022/07/14/reproducing-visualization-in-publications/index.en_files/figure-html/unnamed-chunk-21-1.png" width="672" /></p>
<pre class="r"><code>allGFP %&gt;%
  filter(!is.na(meanCS)) %&gt;%
  mutate(qtile = as.factor(ntile(meanCS, 4))) %&gt;%
  ggplot(aes(x = MatchPos, y = meanCS, color = qtile)) +
  geom_smooth(method = &quot;loess&quot;, span = 0.05, color = &quot;black&quot;, size = 0.5, fill = &quot;grey77&quot;) +
  geom_point(shape = 20) +
  # Use classic theme
  theme_classic() +
  # Remove color legend
  guides(color = &quot;none&quot;) +
  # Set the ticks and limits of x and y axes
  scale_x_continuous(
    limits = c(-50, 750),
    breaks = seq(0, 750, 100),
  ) +
  scale_y_continuous(
    limits = c(-1, 1.5),
    breaks = seq(-1, 1.5, 1),
  ) +
  # Set axis titles
  xlab(&quot;Position in GFP transcript (nt)&quot;) +
  ylab(&quot;log2(FC)&quot;) +
  # Set aspect ratio
  coord_fixed(ratio = 200/3) +
  # Set a customized color palette
  scale_color_manual(values = gfp_color) </code></pre>
<pre><code>## `geom_smooth()` using formula &#39;y ~ x&#39;</code></pre>
<p><img src="https://nyusurp.github.io/2022/07/14/reproducing-visualization-in-publications/index.en_files/figure-html/unnamed-chunk-22-1.png" width="672" /></p>
<p>Many, many annotations are added here. This adds each individual box. Run the following code, and then try changing the Q1 box to say 1Q instead.</p>
<pre class="r"><code>allGFP %&gt;%
  filter(!is.na(meanCS)) %&gt;%
  mutate(qtile = as.factor(ntile(meanCS, 4))) %&gt;%
  ggplot(aes(x = MatchPos, y = meanCS, color = qtile)) +
  geom_smooth(method = &quot;loess&quot;, span = 0.05, color = &quot;black&quot;, size = 0.5, fill = &quot;grey77&quot;) +
  geom_point(shape = 20) +
  # Use classic theme
  theme_classic() +
  # Remove color legend
  guides(color = &quot;none&quot;) +
  # Set the ticks and limits of x and y axes
  scale_x_continuous(
    limits = c(-50, 750),
    breaks = seq(0, 750, 100),
  ) +
  scale_y_continuous(
    limits = c(-1, 1.5),
    breaks = seq(-1, 1.5, 1),
  ) +
  # Set axis titles
  xlab(&quot;Position in GFP transcript (nt)&quot;) +
  ylab(&quot;log2(FC)&quot;) +
  # Set aspect ratio
  coord_fixed(ratio = 200/3) +
  # Set a customized color palette
  scale_color_manual(values = gfp_color) +
  
  
  # Extra annotations -- these are done manually here.
  # The authors wrote a function for this, which is the preferred way:
  # https://gitlab.com/sanjanalab/cas13/-/blob/master/scripts/Plot_GuideScoreDistributionPerGene.R#L140
  # If you really want to fully recapitulate the figure, this is what you need.
  # But manually annotating one figure is pretty tedious as you see below.
  annotate(
    geom = &quot;rect&quot;,
    # xmin and xmax (and x for the text geom should be computed as a function)
    # This is basically the length of each exons in the transcript
    xmin = 0,
    xmax = 750,
    # These are fixed for the transcript plot
    ymin = -1,
    ymax = -0.8,
    color = &quot;grey50&quot;,
    fill = &quot;lightblue&quot;,
    alpha = 0.5
  ) +
  annotate(
    # This labels the exon number
    geom = &quot;text&quot;,
    x = 750/2,
    y = -0.9,
    label = &quot;1&quot;,
    size = 2,
    color = &quot;black&quot;
  ) +
  annotate(
    # The following labels the range of each quartile
    geom = &quot;rect&quot;,
    # The x range is fixed
    xmin = -50,
    xmax = -10,
    # ymin, ymax, and y for the text geom should be computed if a function
    # is to be written
    ymin = min(allGFP$meanCS, na.rm = TRUE),
    ymax = quantile(allGFP$meanCS, 0.25, na.rm = TRUE),
    fill = gfp_color[1],
    alpha = 0.3
  ) +
  annotate(
    geom = &quot;text&quot;,
    x = -30,
    y = 
      (min(allGFP$meanCS, na.rm = TRUE) + quantile(allGFP$meanCS, 0.25, na.rm = TRUE))/2,
    label = &quot;Q1&quot;,
    size = 4
  ) +
  annotate(
    geom = &quot;rect&quot;,
    xmin = -50,
    xmax = -10,
    ymin = quantile(allGFP$meanCS, 0.25, na.rm = TRUE),
    ymax = quantile(allGFP$meanCS, 0.5, na.rm = TRUE),
    fill = gfp_color[2],
    alpha = 0.3
  ) +
  annotate(
    geom = &quot;text&quot;,
    x = -30,
    y = 
      (quantile(allGFP$meanCS, 0.25, na.rm = TRUE) + quantile(allGFP$meanCS, 0.5, na.rm = TRUE))/2,
    label = &quot;Q2&quot;,
    size = 4
  ) +
  annotate(
    geom = &quot;rect&quot;,
    xmin = -50,
    xmax = -10,
    ymin = quantile(allGFP$meanCS, 0.5, na.rm = TRUE),
    ymax = quantile(allGFP$meanCS, 0.75, na.rm = TRUE),
    fill = gfp_color[3],
    alpha = 0.3
  ) +
  annotate(
    geom = &quot;text&quot;,
    x = -30,
    y = 
      (quantile(allGFP$meanCS, 0.5, na.rm = TRUE) + quantile(allGFP$meanCS, 0.75, na.rm = TRUE))/2,
    label = &quot;Q3&quot;,
    size = 4
  ) +
  annotate(
    geom = &quot;rect&quot;,
    xmin = -50,
    xmax = -10,
    ymin = quantile(allGFP$meanCS, 0.75, na.rm = TRUE),
    ymax = max(allGFP$meanCS, na.rm = TRUE),
    fill = gfp_color[4],
    alpha = 0.3
  ) +
  annotate(
    geom = &quot;text&quot;,
    x = -30,
    y = 
      (quantile(allGFP$meanCS, 0.75, na.rm = TRUE) + max(allGFP$meanCS, na.rm = TRUE))/2,
    label = &quot;Q4&quot;,
    size = 4
  )</code></pre>
<pre><code>## `geom_smooth()` using formula &#39;y ~ x&#39;</code></pre>
<p><img src="https://nyusurp.github.io/2022/07/14/reproducing-visualization-in-publications/index.en_files/figure-html/unnamed-chunk-23-1.png" width="672" /></p>
</div>

    </div>
    
    






<div class="share-box" aria-hidden="true">
  <ul class="share">
    
      
      
      
        
      
      
      
      <li>
        <a href="https://twitter.com/intent/tweet?url=https://nyusurp.github.io/2022/07/14/reproducing-visualization-in-publications/&amp;text=Reproducing%20Visualization%20in%20Publications" target="_blank" rel="noopener" class="share-btn-twitter">
          <i class="fab fa-twitter"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="https://www.facebook.com/sharer.php?u=https://nyusurp.github.io/2022/07/14/reproducing-visualization-in-publications/&amp;t=Reproducing%20Visualization%20in%20Publications" target="_blank" rel="noopener" class="share-btn-facebook">
          <i class="fab fa-facebook"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="mailto:?subject=Reproducing%20Visualization%20in%20Publications&amp;body=https://nyusurp.github.io/2022/07/14/reproducing-visualization-in-publications/" target="_blank" rel="noopener" class="share-btn-email">
          <i class="fas fa-envelope"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="https://www.linkedin.com/shareArticle?url=https://nyusurp.github.io/2022/07/14/reproducing-visualization-in-publications/&amp;title=Reproducing%20Visualization%20in%20Publications" target="_blank" rel="noopener" class="share-btn-linkedin">
          <i class="fab fa-linkedin-in"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="whatsapp://send?text=Reproducing%20Visualization%20in%20Publications%20https://nyusurp.github.io/2022/07/14/reproducing-visualization-in-publications/" target="_blank" rel="noopener" class="share-btn-whatsapp">
          <i class="fab fa-whatsapp"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="https://service.weibo.com/share/share.php?url=https://nyusurp.github.io/2022/07/14/reproducing-visualization-in-publications/&amp;title=Reproducing%20Visualization%20in%20Publications" target="_blank" rel="noopener" class="share-btn-weibo">
          <i class="fab fa-weibo"></i>
        </a>
      </li>
    
  </ul>
</div>












  
    
    





  
    
    
    
    
    
    <div class="media author-card content-widget-hr">
      
        
        <img class="avatar mr-3 avatar-square" src="/author/yen-chung-chen/avatar_huda0ce57f65f04f993150b3aa43fca8b8_48438_270x270_fill_q90_lanczos_center.jpg" alt="Yen-Chung Chen">
      

      <div class="media-body">
        <h5 class="card-title"><a href="/author/yen-chung-chen/">Yen-Chung Chen</a></h5>
        <h6 class="card-subtitle">PhD Candidate</h6>
        <p class="card-text">A graduate student interested in developmental biology, neurobiology and bioinformatics.</p>
        <ul class="network-icon" aria-hidden="true">
  
    
    
    
      
    
    
    
    
    
    <li>
      <a href="mailto:ycc520@nyu.edu" >
        <i class="fas fa-envelope"></i>
      </a>
    </li>
  
    
    
    
      
    
    
    
    
    
      
    
    <li>
      <a href="https://twitter.com/ycchen_tw" target="_blank" rel="noopener">
        <i class="fab fa-twitter"></i>
      </a>
    </li>
  
    
    
    
    
    
    
    
      
    
    <li>
      <a href="https://scholar.google.com/citations?user=JtsGxWEAAAAJ" target="_blank" rel="noopener">
        <i class="ai ai-google-scholar"></i>
      </a>
    </li>
  
    
    
    
      
    
    
    
    
    
      
    
    <li>
      <a href="https://github.com/chenyenchung" target="_blank" rel="noopener">
        <i class="fab fa-github"></i>
      </a>
    </li>
  
    
    
    
    
    
    
    
      
    
    <li>
      <a href="https://orcid.org/0000-0002-8529-1251" target="_blank" rel="noopener">
        <i class="ai ai-orcid"></i>
      </a>
    </li>
  
</ul>

      </div>
    </div>
  


  















    
  </div>
</article>

      

    
    
    
      <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha256-9/aliU8dGd2tb6OSsuzixeV4y/faTqgFtohetphbbj0=" crossorigin="anonymous"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.imagesloaded/4.1.4/imagesloaded.pkgd.min.js" integrity="sha256-lqvxZrPLtfffUl2G/e7szqSvPBILGbwmsGE1MKlOi0Q=" crossorigin="anonymous"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.isotope/3.0.6/isotope.pkgd.min.js" integrity="sha256-CBrpuqrMhXwcLLUd5tvQ4euBHCdh7wGlDfNz8vbu/iI=" crossorigin="anonymous"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.5.7/jquery.fancybox.min.js" integrity="sha256-yt2kYMy0w8AbtF89WXb2P1rfjcP/HTHLT7097U8Y5b8=" crossorigin="anonymous"></script>

      

      
        
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.18.1/highlight.min.js" integrity="sha256-eOgo0OtLL4cdq7RdwRUiGKLX9XsIJ7nGhWEKbohmVAQ=" crossorigin="anonymous"></script>
        
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.18.1/languages/r.min.js"></script>
        
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.18.1/languages/bash.min.js"></script>
        
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.18.1/languages/python.min.js"></script>
        
      

    

    
    

    
    
    <script>const code_highlighting = true;</script>
    

    
    
    <script>const isSiteThemeDark = false;</script>
    

    
    
    
    
    
    
    <script>
      const search_config = {"indexURI":"/index.json","minLength":1,"threshold":0.3};
      const i18n = {"no_results":"No results found","placeholder":"Search...","results":"results found"};
      const content_type = {
        'post': "Posts",
        'project': "Projects",
        'publication' : "Publications",
        'talk' : "Talks",
        'slides' : "Slides"
        };
    </script>
    

    
    

    
    
    <script id="search-hit-fuse-template" type="text/x-template">
      <div class="search-hit" id="summary-{{key}}">
      <div class="search-hit-content">
        <div class="search-hit-name">
          <a href="{{relpermalink}}">{{title}}</a>
          <div class="article-metadata search-hit-type">{{type}}</div>
          <p class="search-hit-description">{{snippet}}</p>
        </div>
      </div>
      </div>
    </script>
    

    
    
    <script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/3.2.1/fuse.min.js" integrity="sha256-VzgmKYmhsGNNN4Ph1kMW+BjoYJM2jV5i4IlFoeZA9XI=" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script>
    

    
    

    
    

    
    

    
    
    
    
    
    
    
    
    
      
    
    
    
    
    <script src="/js/academic.min.b2ed8054531fc8b77a9c500d5dfef876.js"></script>

    






  
  
  <div class="container">
    <footer class="site-footer">
  
  <p class="powered-by">
    
      <a href="/privacy/">Privacy Policy</a>
    
    
       &middot; 
      <a href="/terms/">Terms</a>
    
  </p>
  

  <p class="powered-by">
    © NYU 2021 - 2022
  </p>

  
  





  
  
  
  

  
  
    
  
  
    
  

  

  
  <p class="powered-by copyright-license-text">
    This work is licensed under <a href="https://creativecommons.org/licenses/by-nc-sa/4.0" rel="noopener noreferrer" target="_blank">CC BY NC SA 4.0</a>
  </p>
  

  <p class="powered-by footer-license-icons">
    <a href="https://creativecommons.org/licenses/by-nc-sa/4.0" rel="noopener noreferrer" target="_blank">
      <img src="https://search.creativecommons.org/static/img/cc_icon.svg" alt="CC icon">
      <img src="https://search.creativecommons.org/static/img/cc-by_icon.svg" alt="CC by icon">
      
        <img src="https://search.creativecommons.org/static/img/cc-nc_icon.svg" alt="CC NC icon">
      
      
        <img src="https://search.creativecommons.org/static/img/cc-sa_icon.svg" alt="CC SA icon">
      
    </a>
  </p>




  <p class="powered-by">
    Powered by the
    <a href="https://sourcethemes.com/academic/" target="_blank" rel="noopener">Academic theme</a> for
    <a href="https://gohugo.io" target="_blank" rel="noopener">Hugo</a>.

    
    <span class="float-right" aria-hidden="true">
      <a href="#" class="back-to-top">
        <span class="button_icon">
          <i class="fas fa-chevron-up fa-2x"></i>
        </span>
      </a>
    </span>
    
  </p>
</footer>

  </div>
  

  
<div id="modal" class="modal fade" role="dialog">
  <div class="modal-dialog">
    <div class="modal-content">
      <div class="modal-header">
        <h5 class="modal-title">Cite</h5>
        <button type="button" class="close" data-dismiss="modal" aria-label="Close">
          <span aria-hidden="true">&times;</span>
        </button>
      </div>
      <div class="modal-body">
        <pre><code class="tex hljs"></code></pre>
      </div>
      <div class="modal-footer">
        <a class="btn btn-outline-primary my-1 js-copy-cite" href="#" target="_blank">
          <i class="fas fa-copy"></i> Copy
        </a>
        <a class="btn btn-outline-primary my-1 js-download-cite" href="#" target="_blank">
          <i class="fas fa-download"></i> Download
        </a>
        <div id="modal-error"></div>
      </div>
    </div>
  </div>
</div>

</body>
</html>
