---
title: "Workshop 7: Summarizing data"
subtitle: |
    |   - subsetting data
    |   - summarizing your data
    |   - splitting data by categories to calculate stats
    |   - resources for learning R
author:
  - Eugene Plavskin
  - Grace Avecilla
output:
  html_notebook:
    code_folding: show
    depth: 3
    tidy: yes
    toc: yes
---
**Solutions to this workshop can be found [here](Solutions_Workshop_7.nb.html)**

# Review
```{r}
# Read in penguins.csv and save it in a variable called penguins

# Find the range of penguin body mass

# Read in your data and save it as a variable
```

# Subsetting data

One of the most powerful things R can do easily and quickly is to select part of your data based on some criteria you provide using the `subset()` function. This can be immensely useful. Let's say we're interested in penguin data, but don't really care about the *Chinstrap* species Let's create a subset of our data that excludes those animals. To set the conditions for subsetting, you will have to remember something from our lesson on logicals.
```{r}
# create a subset with data on all species *except* setosa
penguin_no_chinstrap <- subset(penguins, species != 'Chinstrap')
```
Look at this new dataframe. Notice that in the code above, subset knows what we mean by `species`; because we already tell is which dataframe to work with, it figures that species is just one of the columns in this dataframe. Recall that `!=` means 'not equal to'.

Try this on your own. Create a data frame of all the penguins where flipper length is greater than 210 mm.
```{r}
# create a subset of the data, penguin_long_flipper, that includes all animals for
# which flipper length is greater than 210 mm

# look at the data frame. what do you notice about these animals (esp their species)?

```

### sidenote: %in%

A really useful logical operator in R is `%in%`; it checks that the thing before it is a member of the list of things after it.
```{r}
# TRUE statement
'a' %in% c('a', 'b', 'c')

# FALSE statement
'e' %in% c('a', 'b', 'c')
```

`%in%` can be incredibly useful in subsetting.
```{r}
# another way to subset data with non-setosa species
penguin_no_chinstrap_2 <- subset(penguins, species %in% c('Adelie', 'Gentoo'))
```

### sidenote: 'and', 'or'

If you want to subset based on multiple logical statements (e.g. 'species is not Chinstrap' AND 'flipper length is greater than 210 mm'), you can use the AND (`&`) and OR (`|`) logical operators. We didn't have time to cover these, but you can still look at our notes on them in the [workshop on logical operations](https://plavskin.github.io/SURP_R_course/Workshop_5.nb.html#operations-on-logicals-and-or).

# Getting useful info about your data

It can be incredibly helpful to quickly get summary info for your data in an accessible, R-readable format. We've already calculated some 'summary statistics' (`mean`) for every column, and we saw how ggplot easily lets us split our plot up by categories (*eg* through using different colors for each one). Now let's do the same with summary statistics.

## Using the summary function

A really useful function for quickly learning about the properties of your data is `summary()`. It shows you the mean, median, and quantiles of your numeric data, as well as counts for each data category. Try this:

```{r}
penguins_summary <- summary(penguins)
print(penguins_summary)
```

One reason `summary()` is often not particularly useful is that it summarizes *all* the data in your dataframe. Notice that it's not splitting the summaries by species, which is probably what we would want to happen.

One way to get around this is through subsetting. Make a subset of the penguin data that only contains data for the *Adelie* species, and then run the `summary` function on this subset.
```{r}
# use subset() to make a dataframe, adelie_df, that contains only the data where
# species is Adelie
# (hint: you want the species column to equal 'Adelie')

# pass the summary data for adelie_df into a new variable called adelie_summary

# print out adelie_summary

```

There are a couple of important limitations to the approach we just used. First of all, we have to do this to one category at a time. This may not be so bad for 3 species, but imagine we had a dataset containing 50 species (or other categories we were interested in). Second, if you look in your **Environment** window, you'll notice that neither penguins_summary or adelie_summary are dataframes. The `summary` function produces a weird R data type called a *table*, which is easy to look at but a pain to do anything with.
```{r}
View(penguins_summary)
```
For example, if you wanted to plot the results of the summary functions for a bunch of species using ggplot (which we'll learn how to do next class!), it would be very annoying to do.

# Summarizing data with plyr

## The problem
Ideally, we want to summarize complicated data in a compact, readable, graphable way, probably as a dataframe. We also don't want to have to subset the data individually for every member of every category we want to split by in our analysis (e.g. subset for every penguin species and then find the mean of some column); this may be simply annoying for 3 penguins, but becomes impossible for many datasets (e.g. imagine trying to get the mean and standard deviation of the expression level of every gene in a single-cell RNAseq experiment).

Although R has built-in functions that do this, we're going to turn to yet another package, **plyr**, which makes this process simpler.

```{r}
# uncomment the line below and install plyr if you haven't already done so
#install.packages('plyr')

# load the plyr library into the current R session
library(plyr)
```

## *split-apply-combine* and basic **plyr** functionality
The basic philosophy behind **plyr** is the idea that for many problems, the solution falls into the pattern of *split-apply-combine*: you want to *split* the data based on some categories, apply a function (usually calculating some kind of summary statistic) to each resulting dataset, and *combine* the results back together.

There are various ways to use **plyr**'s `__ply` functions, and the built-in documentation doesn't offer much help here. If you're interested in diving deeper into this, I can't recommend this tutorial strongly enough:
[**plyr** class from Jenny Bryan's UBC's STAT545 course](https://www.stat.ubc.ca/~jenny/STAT545A/block04_dataAggregation.html)

The application I'm showing you here is the one that makes the most sense to me; it looks something like this:
```{}
output_df <-
  ddply(<input_df>,
        ~ <column names to sort by>,
        summarize,
        <output_column> = <output_function(input_column)>)
```
Here, the `~` represents "as a function of"; this is common to many R functions, as you will see below!

**plyr** command names are based on the data they take as input and output; there are lots of options here, but I don't want to go into too many of them. For now, let's work with the penguin data, and try to get a dataframe out containing some summary stats by species. Since we are going from dataframe (*d*) to dataframe (*d*), the function we want to use is `ddply` (if we were using a dataframe as an output and an array as output, we'd use `daply`, etc).

For example, let's try to create a dataframe that contains the mean bill length of each penguin species:
```{r}
penguin_bill_length_summary <-
  ddply(penguins, ~ species, summarize, mean_bill_Length = mean(bill_length_mm, na.rm = T))
```

## Creating multiple columns of summary data

You can use this format to create multiple new columns using multiple functions; just separate these by commas, like so:
```{r}
penguin_bill_length_summary_2 <-
  ddply(penguins,
        ~ species,
        summarize,
        mean_bill_Length = mean(bill_length_mm, na.rm = T),
        sd_bill_Length = sd(bill_length_mm, na.rm = T))
```
Our summary is now a dataframe, and we can easily retrieve the mean bill length for any species

Let's try a challenge:
```{r}
# Use ddply to create a dataframe, penguin_bill_depth_summary, that contains the
# mean and standard error (SE) of each species' bill depth
# SE = standard deviation / sqrt(count - 1)

```
I think what helps me think about the solution to the above is that, in effect, `ddply` treats each species as its own little dataframe, as if you were manually subsetting the penguins dataframe by species one-by-one and calculating the output column values. This is *split-apply-combine* in action.

## Splitting by multiple categories

Another really great thing about summarizing data this way is that we don't need to limit ourselves to splitting it by a single category. Since the penguin data also includes islands, we can look at island alongside species in our summary statistics.

We can now use ddply just as before, but splitting by a list of categories separated by `+`, rather than by a single category:
```{r}
penguin_bill_length_summary_3 <-
  ddply(penguins,
        ~ species + island,
        summarize,
        mean_bill_Length = mean(bill_length_mm, na.rm = T),
        sd_bill_Length = sd(bill_length_mm, na.rm = T))
```

# Try what you've learned on your own data!
Take some time to explore your data! Try subsetting your data by a category and calculating some summary statistics.

# Things we hope you've learned today (and will hopefully remember next time)

* How to subset data using logical expressions
* How to summarize data in multiple categories using plyr

